File: LeetCode/124. Binary Tree Maximum Path Sum.py
# https://leetcode.com/problems/binary-tree-maximum-path-sum/description/


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxPathSum(self, root: TreeNode | None) -> int:
        max_path = float("-inf")

        def inner(node: TreeNode | None) -> int | float:
            nonlocal max_path

            if node is None:
                return float("-inf")

            left_max_path = inner(node.left)
            right_max_path = inner(node.right)

            #  Выбираем максимальный среди возможных путей в ноде:
            #  1. Сама нода без потомков
            #  2. Путь через ноду и левого потомка
            #  3. Путь через ноду и правого потомка
            through_node_max_path = max(
                node.val, node.val + left_max_path, node.val + right_max_path
            )

            # 4. Плюс путь из левого потомка через ноду в правого потомка (без захода выше в родителя)
            node_max_path = node.val + left_max_path + right_max_path

            max_path = max(max_path, through_node_max_path, node_max_path)

            return through_node_max_path

        inner(root)

        return 0 if isinstance(max_path, float) else max_path

File: LeetCode/124. Binary Tree Maximum Path Sum.js
// https://leetcode.com/problems/binary-tree-maximum-path-sum/description/

class TreeNode {
  constructor(value, left = null, right = null) {
    this.val = value;
    this.left = left;
    this.right = right;
  }
}

const maxPathSum = (root) => {
  let maxPath = Number.NEGATIVE_INFINITY;

  const inner = (node) => {
    if (!node) return Number.NEGATIVE_INFINITY;

    const leftSum = inner(node.left);
    const rightSum = inner(node.right);

    // Выбираем максимальный среди возможных путей в ноде:
    // 1. Сама нода без потомков
    // 2. Путь через ноду и левого потомка
    // 3. Путь через ноду и правого потомка
    const throughNodeMaxPath = Math.max(node.val, node.val + leftSum, node.val + rightSum);

    // 4. Плюс путь из левого потомка через ноду в правого потомка (без захода выше в родителя)
    const nodeMaxPath = node.val + leftSum + rightSum;

    maxPath = Math.max(maxPath, throughNodeMaxPath, nodeMaxPath);

    // Для учета в родителе возвращаем максимальный путь ЧЕРЕЗ текущую ноду
    return throughNodeMaxPath;
  };

  inner(root);

  return maxPath;
};

const root = new TreeNode(-10, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
const root1 = new TreeNode(-10, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));

console.log(maxPathSum(root)); // 42
